UX: Draw/Stroke/Fill better workflow

rodlie opened the pull request:
Based on feedback in https://github.com/orgs/friction2d/discussions/488 I did some changes to the stroke workflow.

* Stroke color and width are now stored for reuse on new draw/stroke action (rectangle/circle)
* Right-click in draw mode will unselect last draw action (or any selected)
  * This allows for changes to color/width during draw session

@pgilfernandez @AptGetMe


https://github.com/user-attachments/assets/a741f7f1-6480-4236-a153-bdf40f771aa9

pgilfernandez commented:
It's working great, I like it!

I would suggest to also include the fill parameter and, finally, could these parameters be remembered or saved as defaults?

AptGetMe commented:
This is fantastic!  🥳️  Exactly what I was hopeful for!  I agree with @pgilfernandez, though.  

> I would suggest to also include the fill parameter...

The same thing for the fill parameter would be icing on the cake 😁️

rodlie commented:
- Linux: https://github.com/friction2d/friction/actions/runs/14420008695
- Windows: https://github.com/friction2d/friction/actions/runs/14420008703
- Mac: https://github.com/friction2d/friction/actions/runs/14420008704
---
- Added support for fill
- Added support for type (fill and/or stroke)
- States are saved during quit (and restored on start)

pgilfernandez commented:
It's working great here, I love it

rodlie commented:
Forgot something, the default fill color is black (always been), this will not work for text (that previous had white hardcoded as default) when the default background is also black.

Should we set the default fill color (for all objects) to white? Else we need to keep track of text fill color also, that complicates things a bit.

pgilfernandez commented:
Umm, the point here is avoiding creating a text and the character not showing up because using the same background color or because the fill color is set to a transparent one (alpha = 0).

I would do some checks to set the color:
- if fill color is transparent then turn it opaque
- if fill color is similar to background color the invert it (is it possible to invert colors?) if not, just check if background color is black and set white for text or the opposite, if background is white set black

Would that solve the issue with text?

rodlie commented:
We should not change the color set by the user, there are cases where the user want's black text regardless of the background color (and what is the background color? what if you have a black background but a red layer over, then the background for the text is actually red). Overriding the user will add more friction ( 😛 )

But we should have a "sane" default for fill and since the default background is black is makes sense (for me) that the default fill is white.

There are of course other edge cases. Example: you add an rectangle, set fill to black and then add a text object and the default is now black, the user might not expect that(?)

One more, if your last stroke color was black, that is now the default, you start Friction several days later and you might be confused that your drawing/circle/rectangle don't show up on your default black background.

rodlie commented:
> One more, if your last stroke color was black, that is now the default, you start Friction several days later and you might be confused that your drawing/circle/rectangle don't show up on your default black background.

Possible fix:

- During startup/new scene compare selected background color against last stroke/fill color and adjust fill/stroke to be visible on background color

rodlie commented:
Updates:

- Added white as default fill color
- On new scene adjust fill/stroke default color against selected background color

pgilfernandez commented:
Yes, nice additions.

BTW, does it make sense to keep last used colors over projects or sessions?

I don't know if it would make sense to reset to default every time we startup Friction...

rodlie commented:
> BTW, does it make sense to keep last used colors over projects or sessions?

Maybe, but that's for another release ;)

> I don't know if it would make sense to reset to default every time we startup Friction...

What do you mean? We now store the last used, that is restored on startup as requested. On new scene we check the last used colors against the background and may adjust last used.

pgilfernandez commented:
Never mind, it's a fantastic step forward so let's keep it as it is now and don't overthink it, hehehe

rodlie commented:
Ok, we can always change/add stuff later. Will merge tomorrow after testing on Windows/Linux.
